### PR TITLE
feat(eslint-plugin): [no-this-alias] report on assignment expressions

### LINKED ISSUE #4690: [no-this-alias] does not detect all cases of aliasing `this` to local variable
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
   "@typescript-eslint/no-this-alias": "error",
  }
}
```

```TS
let that;

that = this;
```

**Expected Result**

3:5  error  Unexpected aliasing of 'this' to local variable  @typescript-eslint/no-this-alias

**Actual Result**

No error or warning.

**Additional Info**

https://github.com/sindresorhus/eslint-plugin-unicorn catches this with unicorn/no-this-assignment

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `15.5.0` |
| `@typescript-eslint/parser`        | `15.5.0` |
| `TypeScript`                       | `4.6.2` |
| `ESLint`                           | `8.9.0` |
| `node`                             | `16.13.2` |

### PR DESCRIPTION
**Pull request by juank1809**
## PR Checklist

-   [x] Addresses an existing issue: fixes #4690 
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

The rule 'no-this-alias' had no logic of how to manage a reassignment of a variable already declared, it was only working in a variable declaration. This code adds an AssignmentExpression node to be handled and grabs the id of the Identifier of either AssigmentExpression or  VariableDeclarator.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-this-alias.ts [hunk 1]
```diff
@@ -48,11 +48,16 @@ export default util.createRule<Options, MessageIds>({
...
+        node: TSESTree.VariableDeclarator | TSESTree.AssignmentExpression,
       ): void {
-        const { id } = node;
+        let id;
 
+        if (node.type === AST_NODE_TYPES.VariableDeclarator) {
+          id = node.id;
+        } else {
+          id = node.left;
+        }
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-this-alias.ts:60]: Just nitpicking, if you wanted to get a little more terse...

```suggestion
        const id =
          node.type === AST_NODE_TYPES.VariableDeclarator ? node.id : node.left;
```

**Review by JoshuaKGoldberg — DISMISSED** [review]
This makes sense, thanks for the change! I'm leaving some comments in the linked issue that might change things, so I'll defer to @bradzacher on merging.

File: packages/eslint-plugin/src/rules/no-this-alias.ts (hunk 1 above)
Comment (juank1809) [reply packages/eslint-plugin/src/rules/no-this-alias.ts:60]: Wow, that definitely seems more readable. I was doing a lot of thinking on how to improve that part of the code, thanks for the feedback!

I must consider using more ternary expressions from now on 😅

I will add a commit to enhance this.

**Review by bradzacher — APPROVED** [review]
this LGTM! Thanks for adding the extra handling!

**Pull request merged**